src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -108,20 +122,80 @@ Score(const size_t queryIndex, TreeType& referenceNode)
 
     // Calculate kernel value based on reference node centroid.
     if (tree::TreeTraits<TreeType>::FirstPointIsCentroid)
-    {
       kernelValue = EvaluateKernel(queryIndex, referenceNode.Point(0));
-    }
     else
-    {
-      kde::KDEStat& referenceStat = referenceNode.Stat();
       kernelValue = EvaluateKernel(queryPoint, referenceStat.Centroid());
-    }
 
     densities(queryIndex) += referenceNode.NumDescendants() * kernelValue;
 
     // Don't explore this tree branch.
     score = DBL_MAX;
   }
+  else if (monteCarlo &&
+           referenceNode.NumDescendants() >= MCAccessCoef * initialSampleSize &&
+           std::is_same<KernelType, kernel::GaussianKernel>::value)
+  {
+    // Calculate reference node depth.
+    if (referenceStat.Depth() == 0)
+      referenceStat.Depth() = CalculateDepth(referenceNode);
```

Comment:
I don't think there's any issue efficiency-wise with putting this code here, but it might be cleaner if you computed the depth during the initialization of `referenceStat`.  You could just do that in the constructor of `KDEStat`.  That could let you remove the code here and also the `CalculateDepth()` function below (or I guess you could move that into `kde_stat.hpp`).

Reply:
I thought about this and I thought I should get the same results. The thing is for ```tree::StandardCoverTree``` I'm getting different depths across reference nodes explored by the single-tree score function, e.g. [diff current implementation (left) vs KDEStat implementation](https://www.diffchecker.com/5WedGrAB) this is specifically from the test ```KDETest/GaussianSingleCoverTreeMonteCarloKDE```. So I thought of making it this way just to be sure depths are right although the code is less clean. In exchange for that you also save time computing depths for query nodes.

Reply:
Wow, that is strange and unexpected.  I wonder if there is a bug in the cover tree implementation.  I don't have the time to debug it right now, but I'll take a look when I get home.  I think for now we should just leave this code as-is.

Reply:
I successfully reproduced this issue with the following code:

```
#include <mlpack/core.hpp>
#include <mlpack/core/tree/cover_tree.hpp>

using namespace mlpack;
using namespace mlpack::tree;
using namespace mlpack::metric;

class DepthStatistic
{
 public:
  DepthStatistic() : depth(0) { }

  template<typename TreeType>
  DepthStatistic(TreeType& node)
  {
    depth = 0;
    TreeType* p = node.Parent();
    while (p != NULL)
    { 
      ++depth;
      p = p->Parent();
    }
  }

  template<typename Archive>
  void serialize(Archive& ar)
  {
    ar & BOOST_SERIALIZATION_NVP(depth);
  }

  const size_t Depth() const { return depth; }

 private:
  size_t depth;
};

int main()
{
  // Create a random dataset and build the tree.
  arma::mat data(2, 3000, arma::fill::randu); // Strangely, the issue does not happen when d=10, n=10k.
  typedef StandardCoverTree<EuclideanDistance, DepthStatistic, arma::mat>
      TreeType;
  TreeType t(std::move(data));

  // Now check each node in the dataset to see if its depth is right.
  std::queue<TreeType*> queue;
  queue.push(&t);

  while (!queue.empty())
  {
    TreeType* node = queue.front();
    queue.pop();

    // Compute depth manually.
    size_t depth = 0;
    TreeType* p = node->Parent();
    while (p != NULL)
    { 
      ++depth;
      p = p->Parent();
    }
    
    if (depth != node->Stat().Depth())
    { 
      Log::Warn << "Depth is different!  We computed " << depth << " now but "
          << node->Stat().Depth() << " when the tree was built!" << std::endl;
    }
    
    for (size_t i = 0; i < node->NumChildren(); ++i)
      queue.push(&node->Child(i));
  }
}
```

I quickly reviewed the cover tree code, and during cover tree construction it is sometimes possible for nodes to be moved around or removed from the tree structure.  Every time this happens, the statistic of a modified child node is rebuilt---but the statistics are not rebuilt recursively!  I bet this is the issue, and tomorrow I'll play with it a bit more and post a patch, which you can just integrate into this PR.

Reply:
Here's a patch you can apply to the branch: https://gist.github.com/071f56263151692ba484e511db859f82

Reply:
Thanks for the patch! :) I found the same problem with ```RectangleTree``` so I solved it in a similar way in 832c423b825a5cb436f0a21c544ed7804e8fa1ec.

Node's depth is now calculated in ```KDEStat``` (432da7bda5d509e706514715e7d086f8d77beefc) and everything seems to work fine.

Reply:
Ah, nice catch, I forgot that that would occur with `RectangleTree` also. :+1: